common/utils/include/claragenomics/utils/buffer.hpp

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
+ *
+ * NVIDIA CORPORATION and its licensors retain all intellectual property
+ * and proprietary rights in and to this software, related documentation
+ * and any modifications thereto.  Any use, reproduction, disclosure or
+ * distribution of this software and related documentation without an express
+ * license agreement from NVIDIA CORPORATION is strictly prohibited.
+ */
+
+#pragma once
+#include <claragenomics/utils/cudautils.hpp>
+#include <claragenomics/utils/allocator.hpp>
+
+namespace claragenomics
+{
+
+template <typename T, typename Allocator>
+class buffer
+{
+public:
+    using value_type     = T;
+    using size_type      = std::size_t;
+    using iterator       = value_type*;
+    using const_iterator = const value_type*;
+
+    buffer(const buffer& other) = delete;
+
+    buffer& operator=(const buffer& other) = delete;
+
+    buffer(size_type n                          = 0,
+           std::shared_ptr<Allocator> allocator = std::make_shared<Allocator>(),
+           cudaStream_t stream                  = 0)
+        : _size(n)
+        , _capacity(n)
+        , _data(nullptr)
+        , _stream(stream)
+        , _allocator(allocator)
+    {
+        if (_capacity > 0)
+        {
+            _data = static_cast<value_type*>(
+                _allocator->allocate(_capacity * sizeof(value_type), _stream));
+            CGA_CU_CHECK_ERR(cudaStreamSynchronize(_stream));
+        }
+    }
+
+    // Should this be removed and force it to use allocator for a common interface?
+    buffer(std::shared_ptr<Allocator> allocator, cudaStream_t stream = 0)
+        : buffer(0, allocator, stream)
+    {
+    }
+
+    // move constructor
+    buffer(buffer&& r)
+        : _size(std::exchange(r._size, 0))
+        , _capacity(std::exchange(r._size, 0))
+        , _data(std::exchange(r._data, nullptr))
+        , _stream(std::exchange(r._stream, cudaStream_t(0)))
+        , _allocator(std::exchange(r._allocator, std::shared_ptr<Allocator>(nullptr)))
+    {
+    }
+
+    ~buffer()
+    {
+        if (nullptr != _data)
+        {
+            _allocator->deallocate(_data, _capacity * sizeof(value_type), _stream);
+        }
+    }
+
+    value_type* data() { return _data; }
+
+    const value_type* data() const { return _data; }
+
+    size_type size() const { return _size; }
+
+    void clear() { _size = 0; }
+
+    iterator begin() { return _data; }
+
+    const_iterator begin() const { return _data; }
+
+    iterator end() { return _data + _size; }
+
+    const_iterator end() const { return _data + _size; }
+
+    void reserve(const size_type new_capacity, cudaStream_t stream = 0)
```

Comment:
Thanks @ahehn-nv. I'm glad you raised this issue, I was debating on this myself.
If we were to make async explicit, then I'd want to make all the member functions as async to be consistent, and this will require every buffer creation to pass a stream. Otherwise, just like you mentioned, we can use the current `stream_` as default for such member functions alone. I'm inclined towards the latter to keep the design simple.
Which makes more sense to you?

Reply:
Yes, I think the latter is fine.

Reply:
done

Reply:
nice solution!